### ISSUE TITLE
OutOfMemoryError when creating bitmaps

### ISSUE THREAD

**Issue by mvglasow** [mapsforge/mapsforge#830]
I just had an unhandled OutOfMemoryError as I resumed SatStat from the background (the app had been in the background for more than 24 hours, the screen had been rotated a few times before, and the position had changed since). I tried to reproduce the error by restarting SatStat, rotating the screen a few times and then rapidly swiping across the map. Indeed it crashed again, albeit with a different call stack.

First crash:

```
SatStat build: git-08592cf


java.lang.OutOfMemoryError: Failed to allocate a 6793310 byte allocation with 3227400 free bytes and 3MB until OOM
	at dalvik.system.VMRuntime.newNonMovableArray(Native Method)
	at android.graphics.Bitmap.nativeCreate(Native Method)
	at android.graphics.Bitmap.createBitmap(Bitmap.java:832)
	at android.graphics.Bitmap.createBitmap(Bitmap.java:809)
	at android.graphics.Bitmap.createBitmap(Bitmap.java:776)
	at org.mapsforge.map.android.graphics.AndroidBitmap.createAndroidBitmap(AndroidBitmap.java:56)
	at org.mapsforge.map.android.graphics.AndroidBitmap.<init>(AndroidBitmap.java:89)
	at org.mapsforge.map.android.graphics.AndroidGraphicFactory.createBitmap(AndroidGraphicFactory.java:197)
	at org.mapsforge.map.view.FrameBuffer.setDimension(FrameBuffer.java:110)
	at org.mapsforge.map.controller.FrameBufferController.onChange(FrameBufferController.java:102)
	at org.mapsforge.map.model.common.Observable.notifyObservers(Observable.java:46)
	at org.mapsforge.map.model.MapViewDimension.setDimension(MapViewDimension.java:34)
	at org.mapsforge.map.android.view.MapView.onSizeChanged(MapView.java:401)
	at android.view.View.sizeChange(View.java:16757)
	at android.view.View.setFrame(View.java:16719)
	at android.view.View.layout(View.java:16636)
	at android.view.ViewGroup.layout(ViewGroup.java:5437)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:336)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:273)
	at android.view.View.layout(View.java:16639)
	at android.view.ViewGroup.layout(ViewGroup.java:5437)
	at android.support.v4.view.ViewPager.onLayout(ViewPager.java:1695)
	at android.view.View.layout(View.java:16639)
	at android.view.ViewGroup.layout(ViewGroup.java:5437)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:336)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:273)
	at android.view.View.layout(View.java:16639)
	at android.view.ViewGroup.layout(ViewGroup.java:5437)
	at android.support.v7.widget.ActionBarOverlayLayout.onLayout(ActionBarOverlayLayout.java:435)
	at android.view.View.layout(View.java:16639)
	at android.view.ViewGroup.layout(ViewGroup.java:5437)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:336)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:273)
	at android.view.View.layout(View.java:16639)
	at android.view.ViewGroup.layout(ViewGroup.java:5437)
	at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1743)
	at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1586)
	at android.widget.LinearLayout.onLayout(LinearLayout.java:1495)
	at android.view.View.layout(View.java:16639)
	at android.view.ViewGroup.layout(ViewGroup.java:5437)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:336)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:273)
	at com.android.internal.policy.PhoneWindow$DecorView.onLayout(PhoneWindow.java:2934)
	at android.view.View.layout(View.java:16639)
	at android.view.ViewGroup.layout(ViewGroup.java:5437)
	at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:2171)
	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:1931)
	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1107)
	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6013)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:858)
	at android.view.Choreographer.doCallbacks(Choreographer.java:670)
	at android.view.Choreographer.doFrame(Choreographer.java:606)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:844)
	at android.os.Handler.handleCallback(Handler.java:739)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:148)
	at android.app.ActivityThread.main(ActivityThread.java:5466)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

Second (provoked) crash:

```
SatStat build: git-08592cf


java.lang.OutOfMemoryError: Failed to allocate a 1179660 byte allocation with 533048 free bytes and 520KB until OOM
	at dalvik.system.VMRuntime.newNonMovableArray(Native Method)
	at android.graphics.BitmapFactory.nativeDecodeStream(Native Method)
	at android.graphics.BitmapFactory.decodeStreamInternal(BitmapFactory.java:635)
	at android.graphics.BitmapFactory.decodeStream(BitmapFactory.java:611)
	at org.mapsforge.map.android.graphics.AndroidTileBitmap.<init>(AndroidTileBitmap.java:114)
	at org.mapsforge.map.android.graphics.AndroidGraphicFactory.createTileBitmap(AndroidGraphicFactory.java:249)
	at org.mapsforge.map.layer.cache.FileSystemTileCache.get(FileSystemTileCache.java:254)
	at org.mapsforge.map.layer.cache.TwoLevelTileCache.setWorkingSet(TwoLevelTileCache.java:101)
	at org.mapsforge.map.layer.TileLayer.draw(TileLayer.java:88)
	at org.mapsforge.map.layer.download.TileDownloadLayer.draw(TileDownloadLayer.java:57)
	at org.mapsforge.map.layer.LayerManager.doWork(LayerManager.java:94)
	at org.mapsforge.map.util.PausableThread.run(PausableThread.java:140)
```

The call stack within Mapsforge is different for both crashes, but both have to do with the creation of bitmaps (either from scratch or from a stream) and both eventually result in a call to `dalvik.system.VMRuntime.newNonMovableArray()`, hence both might be related.

To reproduce, build SatStat as described in https://github.com/mapsforge/mapsforge/issues/649#issuecomment-205483453, using mapsforge c4cc091.

**Comment by devemux86** [comment]
More info would be useful, like Android version, how exactly Mapsforge is used (we show its life cycle in getting started guide and Samples), if label layer is enabled, if overlays are used, etc.

*Also please note that we cannot perform tests with each user's app and its implementation, that's why we keep filling Samples with cases for convenient testing.*

**Comment by mvglasow** [comment]
@devemux86 if I can find the time, I'll investigate myself, hence I just posted the information I had at that point. (And if I do it myself, it'll be easiest to use the app with which I've experienced the issues – that doesn't mean I expect you to debug my app :-)

Android version is 6.0 (CyanogenMod 13). 
Layers used are: Bitmap tile layer, marker and circle overlays (usually 1–2 overlays of each type), no locally rendered tiles or labels.
The activity is configured to respond to orientation changes (default settings), i.e. each rotation will destroy and recreate the activity.

**Comment by devemux86** [comment]
>The activity is configured to respond to orientation changes (default settings), i.e. each rotation will destroy and recreate the activity.

Usually needs good handling of Mapsforge life cycle in that case, also race conditions can play a role.
Why do you prefer that approach and not having the app continue running on orientation changes?

**Comment by mvglasow** [comment]
> Why do you prefer that approach and not having the app continue running on orientation changes?

Not really a design decision, I just stuck with the default. I can look into alternatives – but this might not help with the first crash I've encountered: here the app was in the background for a considerable amount of time (not 100% sure whether or not it was destroyed in the meantime) and crashed when I reopened it.

I see that the "reopen" use case has `android.view.View.sizeChange()` in its call stack, which would presumably get called on rotation if the activity is not destroyed. I thus suspect that keeping the activity on rotation would still crash, but with a different call stack. It would make debugging the first scenario easier, though, if that were the case...

**Comment by mvglasow** [comment]
TL;DR: My bad (at least in part) – but some more documentation (and possibly an exception handler) wouldn't hurt.

I did some debugging by logging the output of `Runtime.getRuntime().totalMemory()` and `Runtime.getRuntime().freeMemory()`, as well as looking at the Samples app.

It turns out total memory used by my app increased steadily with each rotation, until it encountered an `OutOfMemoryError`. Looking at the samples, I saw that the `onDestroy()` method of the activity has calls to `MapView#destroyAll()` and `AndroidGraphicFactory.clearResourceMemoryCache()`. After adding these in my app, the climb in memory usage is a lot less sharp and it didn't crash any more when I tried it – so that seems to have fixed it.

I tried all of this on an experimental branch of my app, based on Mapsforge dev. I then tried to backport the fix to the stable branch (which is based on Mapsforge 0.4.3), only to find out these two calls are not available in that version.

Suggestions:
* Since we've had a couple of API changes, it would help to have a "porting guide" describing what needs to be changed to migrate an app from the previous version to the new one.
* Each removed API call (public method with a particular signature) needs to be documented in the porting guide, including a description how to achieve the same thing in the new version.
* When new API calls become available which must be called in certain situations (such as `MapView#destroyAll()` in the `onDestroy()` method of its parent Activity), document it in the porting guide.
* `AndroidGraphicFactory.clearResourceMemoryCache()` lacks any Javadoc. I added it because the Samples app uses it, but I have no idea what it does or if I really need it – some Javadoc shedding light on the issue would be highly appreciated.
* The two places in which where Mapsforge tries to create a new Android bitmap (either a blank one or from a stream), the possibility of exhausting memory is not taken into account. Is there anything an exception handler could do to handle this situation gracefully (e.g. call  `AndroidGraphicFactory.clearResourceMemoryCache()` or similar to free up some memory)?

**Comment by devemux86** [comment]
We don't generally remove API calls unless there is a very good reason.
All important changes are mentioned in [Changelog](https://github.com/mapsforge/mapsforge/blob/master/docs/Changelog.md).

Until `0.5.1` we used a verbose way to perform cleanup at the exit.
From `0.5.2` we offer a simpler way for that, without removing the low level API.
This is mentioned at `0.5.2-rc1` changelog section as "Simplified cleanup operation #620".

Everyone following Mapsforge guidelines like in Getting Started guide should have already the proper cleanup in the app with one way or another.
For example the change between the two versions can be seen in [0.5.1 guide](https://github.com/mapsforge/mapsforge/blob/0.5.1/docs/Getting-Started-Android-App.md) and the [last one](https://github.com/mapsforge/mapsforge/blob/dev/docs/Getting-Started-Android-App.md).

**Comment by ksaihtam** [comment]
If you work with android studio you can use "Dump Java Heap" before and after rotation. This will help you to identify memory leaks. Cleanup mapsforge on rotation or before application goes into background is a little bit tricky but it is possible.

**Comment by devemux86** [comment]
I suppose we can close this?

**Comment by mvglasow** [comment]
Fine with me. The docs are there by now – my buggy code might have been older than the docs, but I have since RTFM'ed and cleaned up my code, no more crashes :-)
